engine/table/src/main/java/io/deephaven/engine/table/impl/select/StatefulSelectColumn.java

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.engine.table.impl.select;
+
+import io.deephaven.annotations.SimpleStyle;
+import io.deephaven.engine.rowset.TrackingRowSet;
+import io.deephaven.engine.table.ColumnDefinition;
+import io.deephaven.engine.table.ColumnSource;
+import io.deephaven.engine.table.WritableColumnSource;
+import io.deephaven.engine.table.impl.MatchPair;
+import org.immutables.value.Value;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link SelectColumn} that wraps another {@code SelectColumn}, and is used to indicate that the wrapped column has
+ * side effects that prevent parallelization. (e.g. a column that depends on processing rows in order, or would suffer
+ * from lock contention if parallelized)
+ */
+@Value.Immutable
+@SimpleStyle
+public abstract class StatefulSelectColumn implements SelectColumn {
+
+    public static StatefulSelectColumn of(@NotNull final SelectColumn innerSelectColumn) {
+        return ImmutableStatefulSelectColumn.of(innerSelectColumn);
```

Comment:
instanceof to return it if it's already stateful? Or, what if `innerSelectColumn.isStateless() == true`? We might decide to widen return type to just `SelectColumn`?

I don't know if the contract is you can only call isStateless after initInputs (ie, the isStateless call itself is stateful, depending on setup); in which case wrapping as you have done is the correct way to force regardless.